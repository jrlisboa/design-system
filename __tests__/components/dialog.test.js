import React from 'react';
import ReactDOM from 'react-dom';

import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react'
import { expect, jest } from '@jest/globals';

import 'jest-styled-components'

import Dialog from '../../src/components/dialog';

describe('Dialog test', () => {
    describe('Component setup', () => {
        beforeAll(() => {
            ReactDOM.createPortal = jest.fn((element, _) => element);
        });

        afterEach(() => {
            ReactDOM.createPortal.mockClear();
        });

        it('creates a portal root when render', async () => {
            render(
                <Dialog
                    isOpen
                    onClose={() => { }}
                    closeOnOverlayClick />,
            );
            const portalRoot = document.getElementById('dialog-root');
            expect(portalRoot.outerHTML).toBe(`<div id="dialog-root"></div>`);
        })

        it('should return null when isOpen is false', async () => {
            const component = renderer.create(
                <Dialog
                    isOpen={false}
                    onClose={() => { }}
                    closeOnOverlayClick={false} />,
            );
            expect(component.toJSON()).toMatchSnapshot();
            expect(component.toJSON()).toBe(null);
        })

        it('verify if dialog has all expected properties', async () => {
            const { getByTestId } = render(
                <Dialog
                    isOpen
                    title="Title"
                    onClose={() => { }}
                    closeOnOverlayClick>
                    <p>Content</p>
                </Dialog>
            );

            expect(getByTestId('DialogWrapper')).toBeTruthy()
            expect(getByTestId('Dialog')).toBeTruthy()
            expect(getByTestId('DialogHeader')).toBeTruthy()
            expect(getByTestId('DialogTitle')).toBeTruthy()
            expect(getByTestId('DialogCloseButton')).toBeTruthy()
            expect(getByTestId('DialogContent')).toBeTruthy()
        })
    })

    describe('Component useEffect behaviors', () => {
        let events = {};

        beforeEach(() => {
            events = {};
            document.addEventListener = jest.fn((event, callback) => {
                events[event] = callback;
            });

            document.removeEventListener = jest.fn((event, _) => {
                delete events[event];
            });
        });

        beforeAll(() => {
            ReactDOM.createPortal = jest.fn((element, _) => element);
        });

        afterEach(() => {
            ReactDOM.createPortal.mockClear();
            document.body.style.overflow = '';
        });

        it('should set body overflow hidden when isOpen is true', async () => {
            render(
                <Dialog
                    isOpen
                    onClose={() => { }}
                    closeOnOverlayClick={false} />,
            );
            expect(document.body.style.overflow).toBe('hidden');
        })

        it('should change body overflow to unset when isOpen is false', async () => {
            const { rerender } = render(
                <Dialog
                    isOpen
                    onClose={() => { }}
                    closeOnOverlayClick={false} />,
            );
            expect(document.body.style.overflow).toBe('hidden');

            rerender(
                <Dialog
                    isOpen={false}
                    onClose={() => { }}
                    closeOnOverlayClick={false} />,
            );
            expect(document.body.style.overflow).toBe('unset');
        })

        it('should set body overflow unset when component will unmount', async () => {
            const { unmount } = render(
                <Dialog
                    isOpen
                    onClose={() => { }}
                    closeOnOverlayClick={false} />,
            );
            expect(document.body.style.overflow).toBe('hidden');

            unmount();
            expect(document.body.style.overflow).toBe('unset');
        })

        it('calls onClose action when Escape key is pressed', async () => {
            const handleProp = { close: () => { } };
            render(
                <Dialog
                    isOpen
                    onClose={() => handleProp.close()}
                    closeOnOverlayClick />,
            );
            jest.spyOn(handleProp, 'close');
            events.keydown({ key: 'Escape', preventDefault: () => { } });

            expect(handleProp.close).toBeCalledTimes(1);
        })

        it('calls onClose and remove event listener when component will unmount', async () => {
            const handleProp = { close: () => { } };
            const { unmount } = render(
                <Dialog
                    isOpen
                    onClose={() => handleProp.close()}
                    closeOnOverlayClick />,
            );

            jest.spyOn(handleProp, 'close');
            unmount();

            expect(handleProp.close).toBeCalledTimes(1);
            expect(Object.keys(events).length).toBe(0);
        })
    })

    describe('Component snapshot check', () => {
        it('renders correctly', async () => {
            const component = renderer.create(
                <Dialog
                    isOpen
                    title="Title"
                    onClose={() => { }}
                    closeOnOverlayClick>
                    <p>Content</p>
                </Dialog>
            ).toJSON();
            expect(component).toMatchSnapshot();
        })

        it('adds another child item', async () => {
            const component = renderer.create(
                <Dialog
                    isOpen
                    title="Title"
                    onClose={() => { }}
                    closeOnOverlayClick>
                    <p>Content</p>
                    <p>Content 2</p>
                </Dialog>
            ).toJSON();
            expect(component).toMatchSnapshot();
        })

        it('changes closeOnOverlayClick attribute', async () => {
            const component = renderer.create(
                <Dialog
                    isOpen
                    title="Title"
                    onClose={() => { }}
                    closeOnOverlayClick={false}>
                    <p>Content</p>
                </Dialog>
            ).toJSON();
            expect(component).toMatchSnapshot();
        })

        it('changes isOpen attribute', async () => {
            const component = renderer.create(
                <Dialog
                    isOpen={false}
                    title="Title"
                    onClose={() => { }}
                    closeOnOverlayClick>
                    <p>Content</p>
                </Dialog>
            ).toJSON();
            expect(component).toMatchSnapshot();
        })

        it('render just with required attributes', async () => {
            const component = renderer.create(
                <Dialog
                    isOpen={false}
                    onClose={() => { }}
                    closeOnOverlayClick />
            ).toJSON();
            expect(component).toMatchSnapshot();
        })
    })

    describe('Component click behaviors', () => {
        it('closes dialog on background layer click', async () => {
            const handleProp = { close: () => { } };
            const { getByTestId } = render(
                <Dialog
                    isOpen
                    title="Title"
                    onClose={() => handleProp.close()}
                    closeOnOverlayClick>
                    <p>Content</p>
                </Dialog>
            );
            jest.spyOn(handleProp, 'close');
            fireEvent.click(getByTestId('DialogWrapper'));

            expect(handleProp.close).toBeCalledTimes(1);
        })

        it('ignores click on background layer', async () => {
            const handleProp = { close: () => { } };
            const { getByTestId } = render(
                <Dialog
                    isOpen
                    title="Title"
                    onClose={() => handleProp.close()}
                    closeOnOverlayClick={false}>
                    <p>Content</p>
                </Dialog>
            );
            jest.spyOn(handleProp, 'close');
            fireEvent.click(getByTestId('DialogWrapper'));

            expect(handleProp.close).toBeCalledTimes(0);
        })

        it('closes dialog with DialogCloseButton click', async () => {
            const handleProp = { close: () => { } };
            const { getByTestId } = render(
                <Dialog
                    isOpen
                    title="Title"
                    onClose={() => handleProp.close()}
                    closeOnOverlayClick>
                    <p>Content</p>
                </Dialog>
            );
            jest.spyOn(handleProp, 'close');
            fireEvent.click(getByTestId('DialogCloseButton'));

            expect(handleProp.close).toBeCalledTimes(1);
        })
    })
})