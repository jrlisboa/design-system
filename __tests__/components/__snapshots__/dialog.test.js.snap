// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog test Component setup should return null when isOpen is false 1`] = `null`;

exports[`Dialog test Component snapshot check adds another child item 1`] = `
.c0 {
  position: fixed;
  inset: 0;
  width: 100%;
  height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  width: 100%;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  padding: 0;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  height: 100%;
  width: 32px;
}

.c3:hover {
  cursor: pointer;
}

.c4 {
  max-height: calc(80vh - );
  min-height: 20vh;
  overflow-y: auto;
}

.c4 img {
  max-width: 100%;
  object-fit: cover;
}

@media (min-width:600px) {
  .c0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

@media (min-width:600px) {
  .c1 {
    width: 50%;
    -webkit-align-self: center;
    -ms-flex-item-align: center;
    align-self: center;
  }
}

<div
  className="c0"
  data-testid="DialogWrapper"
  onClick={[Function]}
>
  <div
    className="c1"
    data-testid="Dialog"
    onClick={[Function]}
  >
    <div
      className="c2"
      data-testid="DialogHeader"
    >
      <h2
        className=""
        data-testid="DialogTitle"
      >
        Title
      </h2>
      <div
        className="c3"
        data-testid="DialogCloseButton"
        onClick={[Function]}
      >
        <img
          alt="Close"
        />
      </div>
    </div>
    <div
      className="c4"
      data-testid="DialogContent"
    >
      <p>
        Content
      </p>
      <p>
        Content 2
      </p>
    </div>
  </div>
</div>
`;

exports[`Dialog test Component snapshot check changes closeOnOverlayClick attribute 1`] = `
.c0 {
  position: fixed;
  inset: 0;
  width: 100%;
  height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  width: 100%;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  padding: 0;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  height: 100%;
  width: 32px;
}

.c3:hover {
  cursor: pointer;
}

.c4 {
  max-height: calc(80vh - );
  min-height: 20vh;
  overflow-y: auto;
}

.c4 img {
  max-width: 100%;
  object-fit: cover;
}

@media (min-width:600px) {
  .c0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

@media (min-width:600px) {
  .c1 {
    width: 50%;
    -webkit-align-self: center;
    -ms-flex-item-align: center;
    align-self: center;
  }
}

<div
  className="c0"
  data-testid="DialogWrapper"
>
  <div
    className="c1"
    data-testid="Dialog"
    onClick={[Function]}
  >
    <div
      className="c2"
      data-testid="DialogHeader"
    >
      <h2
        className=""
        data-testid="DialogTitle"
      >
        Title
      </h2>
      <div
        className="c3"
        data-testid="DialogCloseButton"
        onClick={[Function]}
      >
        <img
          alt="Close"
        />
      </div>
    </div>
    <div
      className="c4"
      data-testid="DialogContent"
    >
      <p>
        Content
      </p>
    </div>
  </div>
</div>
`;

exports[`Dialog test Component snapshot check changes isOpen attribute 1`] = `null`;

exports[`Dialog test Component snapshot check render just with required attributes 1`] = `null`;

exports[`Dialog test Component snapshot check renders correctly 1`] = `
.c0 {
  position: fixed;
  inset: 0;
  width: 100%;
  height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  width: 100%;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  padding: 0;
}

.c2 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  height: 100%;
  width: 32px;
}

.c3:hover {
  cursor: pointer;
}

.c4 {
  max-height: calc(80vh - );
  min-height: 20vh;
  overflow-y: auto;
}

.c4 img {
  max-width: 100%;
  object-fit: cover;
}

@media (min-width:600px) {
  .c0 {
    -webkit-box-pack: center;
    -webkit-justify-content: center;
    -ms-flex-pack: center;
    justify-content: center;
  }
}

@media (min-width:600px) {
  .c1 {
    width: 50%;
    -webkit-align-self: center;
    -ms-flex-item-align: center;
    align-self: center;
  }
}

<div
  className="c0"
  data-testid="DialogWrapper"
  onClick={[Function]}
>
  <div
    className="c1"
    data-testid="Dialog"
    onClick={[Function]}
  >
    <div
      className="c2"
      data-testid="DialogHeader"
    >
      <h2
        className=""
        data-testid="DialogTitle"
      >
        Title
      </h2>
      <div
        className="c3"
        data-testid="DialogCloseButton"
        onClick={[Function]}
      >
        <img
          alt="Close"
        />
      </div>
    </div>
    <div
      className="c4"
      data-testid="DialogContent"
    >
      <p>
        Content
      </p>
    </div>
  </div>
</div>
`;
